File: app/src/main/java/org/kindone/willingtodo/pomodorotimer/TimerRunning.java
package org.kindone.willingtodo.pomodorotimer;

/**
 * Created by kindone on 2017. 1. 1..
 */

public class TimerRunning extends TimerState {

    final private long mDurationMs;
    final private long mStartedMs;

    public TimerRunning(TimerContext context, long durationMs) {
        super(context);
        mStartedMs = System.currentTimeMillis();
        mDurationMs = durationMs;
    }

    public TimerRunning(TimerContext context, long durationMs, long startedMs) {
        super(context);
        mStartedMs = startedMs;
        mDurationMs = durationMs;
    }

    @Override
    public boolean pause() {
        long elapsedMs = System.currentTimeMillis() - mStartedMs;
        long remainingMs = mDurationMs - elapsedMs;
        changeState(new TimerPaused(mContext, remainingMs));
        return CHANGED;
    }

    @Override
    public boolean start() {
        // DO NOTHING
        return UNCHANGED;
    }

    @Override
    public boolean resume() {
        // DO NOTHING
        return UNCHANGED;
    }

    @Override
    public boolean stop() {
        changeState(new TimerStopped(mContext));
        return CHANGED;
    }

    @Override
    public long getRemainingTimeMs() {
        long elapsedMs = System.currentTimeMillis() - mStartedMs;
        long remainingMs = mDurationMs - elapsedMs;
        return remainingMs > 0 ? remainingMs : 0;
    }

    public long getStartedTimeMs() { return mStartedMs; }

    public boolean isRunning() {
        return true;
    }

    public boolean isStopped() {
        return false;
    }
}

File: app/src/main/java/org/kindone/willingtodo/pomodorotimer/TimerRunning.kt
package org.kindone.willingtodo.pomodorotimer

/**
 * Created by kindone on 2017. 1. 1..
 */

class TimerRunning : TimerState {

    private val mDurationMs: Long
    private val mStartedMs: Long

    constructor(context: TimerContext, durationMs: Long) : super(context) {
        mStartedMs = System.currentTimeMillis()
        mDurationMs = durationMs
    }

    constructor(context: TimerContext, durationMs: Long, startedMs: Long) : super(context) {
        mStartedMs = startedMs
        mDurationMs = durationMs
    }

    override fun pause(): Boolean {
        val elapsedMs = System.currentTimeMillis() - mStartedMs
        val remainingMs = mDurationMs - elapsedMs
        changeState(TimerPaused(mContext, remainingMs))
        return TimerState.CHANGED
    }

    override fun start(): Boolean {
        // DO NOTHING
        return TimerState.UNCHANGED
    }

    override fun resume(): Boolean {
        // DO NOTHING
        return TimerState.UNCHANGED
    }

    override fun stop(): Boolean {
        changeState(TimerStopped(mContext))
        return TimerState.CHANGED
    }

    override fun getRemainingTimeMs(): Long {
        val elapsedMs = System.currentTimeMillis() - mStartedMs
        val remainingMs = mDurationMs - elapsedMs
        return if (remainingMs > 0) remainingMs else 0
    }

    override fun getStartedTimeMs(): Long {
        return mStartedMs
    }

    override fun isRunning(): Boolean {
        return true
    }

    override fun isStopped(): Boolean {
        return false
    }
}
